### ISSUE TITLE
Incompatible type with `many=True`

### ISSUE THREAD

**Issue by christianbundy** [typeddjango/djangorestframework-stubs#260]
# Bug report

## What's wrong

```python3
class MySerializer(serializers.Serializer[int]): ...

MySerializer([1,2,3], many=True)
```

```
error: Argument 1 to "MySerializer" has incompatible type "List[int]"; expected "Optional[int]"  [arg-type]
```

## How is that should be

```
Success: no issues found in 1 source file
```

## System information

- OS: `Linux f16759b1b324 5.10.104-linuxkit #1 SMP PREEMPT Thu Mar 17 17:05:54 UTC 2022 aarch64 GNU/Linux`
- `python` version: 3.10.6
- `django` version: 4.0.7
- `mypy` version: 0.961
- `django-stubs` version: 1.12.0

Also, it isn't asked for by the template, but:

- `djangorestframework` version: 3.13.1
- `djangorestframework-stubs` version: 1.7.0

**Comment by mdantonio** [comment]
We are observing exactly the same error, in our case we have:

- `python` 3.9.13
- `django` 4.1.1
- `mypy` 0.961
- `django-stubs` version: 1.12.0
- `djangorestframework` version: 3.13.1
- `djangorestframework-stubs` version: 1.7.0

I see that BaseSerializer.instance is annotated as:

```
    instance: Optional[_IN]
```

That explains why the list is not accepted, even if it is a correct DRF syntax when passing many=True

**Comment by sshishov** [comment]
I am having the issue with `many=True` as well, but actually the problem is the type of the instance when we are calling save method (`instances = serializer.save()`) when serializer was initialized with `many=True`.

The instance type is returned `Instance` instead of `list[Instance]` or `Sequence` or `Iterable`...

**Comment by Viicos** [comment]
This looks tricky to fix. When `many=True` is used, the `many_init` classmethod is used to return a `ListSerializer` instance. So first of all this method should return `ListSerializer` instead of `BaseSerializer`. Assuming people specifying a [custom class](https://github.com/typeddjango/djangorestframework-stubs/blob/728da9c1ba7fa2a6eb3442eb599d0b38e6529492/rest_framework-stubs/serializers.pyi#L108) are inheriting from `ListSerializer`, this should be ok.
https://github.com/typeddjango/djangorestframework-stubs/blob/728da9c1ba7fa2a6eb3442eb599d0b38e6529492/rest_framework-stubs/serializers.pyi#L108

Then an overload should be used here:
https://github.com/typeddjango/djangorestframework-stubs/blob/728da9c1ba7fa2a6eb3442eb599d0b38e6529492/rest_framework-stubs/serializers.pyi#L84

To return something `ListSerializer` instead of `Self` if `many=True`.

Finally, an overload similar to the previous one should be used in the `__init__` method:

```python
@overload
def __init__(
    self,
    instance: List[_IN] | None = ...,
    data: Any = ...,
    partial: Literal[True] = ...,
    many: bool = ...,
    ...
```

But type checkers would probably not be able to infer the returned instance based on the `many` argument anyway.

**Comment by christianbundy** [comment]
> But type checkers would probably not be able to infer the returned instance based on the many argument anyway.

I was imagining that we might be able to use an overload here, discriminated on whether `many: Literal[True]` or `many: Literal[False]` -- do you think that would work, or are there gotchas?

**Comment by Viicos** [comment]
> > But type checkers would probably not be able to infer the returned instance based on the many argument anyway.
> 
> I was imagining that we might be able to use an overload here, discriminated on whether `many: Literal[True]` or `many: Literal[False]` -- do you think that would work, or are there gotchas?

Made a first draft in the linked PR, I'd be happy to have your insight on it as I'm unsure this would work :)

**Comment by sshishov** [comment]
I assume this is the related issue on the main repo: https://github.com/encode/django-rest-framework/issues/8926

**Comment by flaeppe** [comment]
Would it be possible to `@overload` on `__new__`? To capture returning a `ListSerializer`? As mypy should consider `__new__`s return type nowadays

**Comment by Viicos** [comment]
> Would it be possible to `@overload` on `__new__`? To capture returning a `ListSerializer`? As mypy should consider `__new__`s return type nowadays

You can check the pending PR https://github.com/typeddjango/djangorestframework-stubs/pull/315, `mypy` doesn't allow all `__new__`s return type unfortunately

**Comment by flaeppe** [comment]
Right, I suppose the simplest way to get types right is to just call `.many_init(...)` directly instead of passing `many=True` via `__new__`

**Comment by PedroPerpetua** [comment]
Has this issue hit a wall? I tried to follow he "issue trail" and it seems it's open issues left and right.

The ability to have `serializer.data` return a `list` if `many=True` would be specially useful in testing when comparing API responses on list endpoints with expected data; `assertListEqual` provides a lot more useful feedback than `assertEqual`, but mypy will complain that `serializer.data` is not a `list`.

**Comment by Viicos** [comment]
We need https://github.com/python/mypy/pull/16020.
